frontend/app/components/error-summary.tsx

```diff
@@ -1,48 +1,248 @@
+ *   and focuses on it to ensure visibility.
+ * - Provides an accessible, styled list of errors with links to the form fields.
+ *
+ * @returns The rendered component or `false` if there are no errors.
+ */
+export function ErrorSummary({ className, errors, ...rest }: ErrorSummaryProps): JSX.Element | false {
   const { t } = useTranslation(['gcweb']);
+  const sectionRef = useRef<HTMLElement>(null);
+
+  // Scroll and focus on the error summary when errors are updated.
+  useEffect(() => {
+    if (errors.length > 0 && sectionRef.current) {
+      sectionRef.current.scrollIntoView({ behavior: 'smooth' });
+      sectionRef.current.focus();
+    }
+  }, [errors]);
+
+  // If no errors, return false (no rendering).
+  if (errors.length === 0) {
+    return false;
+  }
+
+  return (
+    <section
+      ref={sectionRef}
+      className={cn('my-5 border-4 border-red-600 p-4', className)}
+      {...rest}
+      tabIndex={-1}
+      role="alert"
+    >
+      <h2 className="font-lato text-lg font-semibold">{t('gcweb:error-summary.header', { count: errors.length })}</h2>
+      <ol className="mt-1.5 list-decimal space-y-2 pl-7">
+        {errors.map(({ fieldId, errorMessage }) => (
+          <li key={`${fieldId}-${errorMessage}`}>
+            <AnchorLink className="text-red-700 underline hover:decoration-2 focus:decoration-2" anchorElementId={fieldId}>
+              {errorMessage}
+            </AnchorLink>
+          </li>
+        ))}
+      </ol>
+    </section>
+  );
+}
+
+/**
+ * Props for the `ActionDataErrorSummary` component.
+ */
+type ActionDataErrorSummaryProps = ComponentPropsWithoutRef<'div'> & {
+  /**
+   * The action data from a React Router Form or Fetcher.
+   * This triggers error collection when form or Fetcher data changes.
+   */
+  actionData: unknown | undefined;
+};
+
+/**
+ * The `ActionDataErrorSummary` component collects and displays form validation errors.
+ *
+ * @remarks
+ * This component uses the `useEffect` hook to capture validation errors from form inputs that contain
+ * the `aria-errormessage` attribute. It finds the error messages, associates them with input fields,
+ * and displays them using the `ErrorSummary` component.
+ *
+ * @example
+ * ```tsx
+ * import { Form } from 'react-router';
+ * import type { Route } from './+types/my-route';
+ * import { ActionDataErrorSummary } from '~/components/error-summary';
+ *
+ * export default function MyRouteComponent({ loaderData, actionData, params }: Route.ComponentProps) {
+ *   return (
+ *     <ActionDataErrorSummary actionData={actionData}>
+ *       <Form method="post" noValidate>
+ *         ...
+ *       </Form>
+ *     </ActionDataErrorSummary>
+ *   );
+ * }
+ * ```
+ */
+export function ActionDataErrorSummary({ actionData, children, ...rest }: ActionDataErrorSummaryProps): JSX.Element {
+  const [errors, setErrors] = useState<ErrorSummaryItem[]>([]);
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  useEffect(() => {
```

Comment:
Might be a nitpick but it could be helpful to extract the code in this `useEffect()` and name it something that describe what it does. It's not easy to glance at this code and know what it's doing.. there's not even a comment explaining why this `useEffect()` is necessary.

Reply:
Read component's [@remark](https://github.com/DTS-STN/future-sir/pull/194/files#diff-82595ba5d71248d5d4408631712369e625228ab33d6a97e3ba1aecf381d0fe43R93-R96)

Reply:
I still think the code is hard to read. How are devs supposed to know they have to scroll up almost 20 lines to see what this block of 30 lines of code does.. I obviously didn't.

I think this code would benefit greatly if it were broken into well named functions (or even a single function).

Reply:
It's been broken into well named functions as requested.